pool.ts

```diff
@@ -0,0 +1,109 @@
+import { dial } from "deno";
+import { Client } from "./client.ts";
+import { Connection } from "./connection.ts";
+import { ConnectionParams, IConnectionParams } from "./connection_params.ts";
+import { Query, QueryConfig, QueryResult } from "./query.ts";
+import { defer, Deferred } from "./deps.ts";
+
+export class Pool {
+  private _connectionParams: IConnectionParams;
+  private _connections: Array<Connection>;
+  private _availableConnections: DeferredStack<Connection>;
+  private _size: number;
+  private _ready: Promise<void>;
+
+  constructor(connectionParams: IConnectionParams, size: number) {
+    this._connectionParams = connectionParams;
+    this._size = size;
+    this._ready = this.startup();
+  }
+
+  private async newConnection(): Promise<Connection> {
+    const { host, port } = this._connectionParams;
+    let addr = `${host}:${port}`;
+
+    const conn = await dial("tcp", addr);
+    const connection = new Connection(conn, this._connectionParams);
+
+    await connection.startup({ ...this._connectionParams });
+    await connection.initSQL();
+    return connection;
+  }
+
+  get size(): number {
+    return this._size;
+  }
+  get available(): number {
+    return this._availableConnections.size;
+  }
+
+  private async startup(): Promise<void> {
+    const connecting = [...Array(this.size)].map(
+      async () => await this.newConnection()
+    );
+    this._connections = await Promise.all(connecting);
+    this._availableConnections = new DeferredStack(this._connections);
+  }
+
+  async end(): Promise<void> {
+    await this._ready;
+    const ending = this._connections.map(c => c.end());
+    await Promise.all(ending);
+  }
+
+  // TODO: can we use more specific type for args?
+  async query(
+    text: string | QueryConfig,
+    ...args: any[]
+  ): Promise<QueryResult> {
+    const query = new Query(text, ...args);
+    return await this.execute(query);
+  }
+
+  private async execute(query: Query): Promise<QueryResult> {
+    await this._ready;
+    const connection = await this._availableConnections.pop();
+    const result = await query.execute(connection);
+    this._availableConnections.push(connection);
+    return result;
+  }
+
+  async connect(): Promise<Client> {
+    await this._ready;
```

Comment:
Call startup here first time it's accessed, and set `_aenter = this.connect`

Reply:
Good point. The node-postgres API is a little weird that it does this implictly, my preference would be to connect asap (so you don't have to wait on the first query, but meh)

Reply:
Actually I like that `node-postgres` creates clients in pool lazily, but that's for another PR

Reply:
Much better to fail (if it can't make N connections) at boot-time, before the instance is attached to a load-balancer, than to fail at run-time. Better for perf too to keep N connections continuously open.

This should, at least, be an option (I'd argue it should be the default).

Reply:
Agreed.